_posts/2020-03-29-all-you-need-is-lambda-1-booleans.md

```diff
@@ -0,0 +1,260 @@
+
+
+```
+if = λ cond then else . cond then else
+```
+
+> Aside: under standard semantics for the lambda calculus, we could simplify this definition further by “η-reducing” it, noting that `λ x . f x` behaves the same as `f` alone. Nevertheless, we will leave this and likely other definitions in their full, η-long forms for the sake of clarity.
+
+This feels strangely like cheating: surely we’ve only moved the problem around. Now instead of `if` making the decision about which argument to return, we’ve deferred it to `cond`. But `if` and `cond` aren’t the same kind of thing, semantically; `if` takes a boolean and two other arguments and returns one of the latter, while `cond` _is_ a boolean—albeit evidently a boolean represented as a function. Let’s make that precise by writing down `if`’s type:
+
+```
+if : Bool -> a -> a -> a
+```
+
+Notwithstanding our use of the yet-to-be-defined name `Bool` for the type of the condition, this is the same type as we gave `if_` in Haskell; that’s a good sign that we’re on the right track! It takes a `Bool` and two arguments of type `a`, and it must return one of those because that’s the only way for it to come up with the `a` that it returns. But what _is_ `Bool`?
+
+Working backwards from the type and definition of `if`, we see that `cond` is applied to two arguments, and therefore must be a function of two parameters. Further, these are both of type `a`, and the value it returns must also be of type `a` for `if`’s type to hold. Thus, we can define the type `Bool` like so:
+
+```
+Bool = ∀ a . a -> a -> a
+```
+
+> Aside: I’m making explicit use of the for-all quantifier here to drive home the point that any particular `Bool` value must be able to be applied to `then` and `else` values of any arbitrary type `a`, defined now or in the future.
+>
+> By the same token, we could have written `if`’s type more explicitly as:
+>
+> ```
+> if : ∀ a . Bool -> a -> a -> a
+> ```
+>
+> Here and in future, local type variables can be assumed to be implicitly generalized in the same manner as Haskell if not otherwise quantified.
+
+If a given `Bool` is a function of two arguments of arbitrary type, returning the same type, it must therefore select one of its arguments to return. There are only two distinguishable inhabitants of `Bool`, `true` and `false`, so we can therefore deduce that since `if` defers the selection of the result to the `Bool`, for `true` and `false` to actually differ they must make opposite selections. In other words, `true` must return the `then` parameter, while `false` must return the `else` one:
+
+
+```
+true, false : Bool
+true  = λ then else . then
+false = λ then else . else
+```
+
+We didn’t move the problem around at all—we actually solved it (once we defined `true` and `false`, anyway). It’s simply that encoding booleans in this manner makes `if` largely pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.
```

Comment:
```suggestion
We didn’t move the problem around at all—we solved it (once we defined `true` and `false`, anyway). The fact is that encoding booleans in this manner makes `if` pointless, since if we can apply `if` to a `Bool` and two arguments, we could equally apply the `Bool` to those arguments directly.
```